### PR TITLE
feat: lsp definitions

### PR DESCRIPTION
**Pull request by oberblastmeister**
checkout [this](https://github.com/nvim-telescope/telescope.nvim/issues/629) issue for more info

### CODE REVIEW

File: lua/telescope/builtin/lsp.lua [hunk 1]
```diff
@@ -38,6 +38,39 @@ lsp.references = function(opts)
...
+      prompt_title = 'LSP Definitions',
+      finder = finders.new_table {
+        results = locations,
+        entry_maker = opts.entry_maker or make_entry.gen_from_quickfix(opts),
+      },
+      previewer = conf.qflist_previewer(opts),
+      sorter = conf.generic_sorter(opts),
+    }):find()
+  end
+end
```
Comment (jaywonchung) [inline lua/telescope/builtin/lsp.lua:72]:
```suggestion
  if #flattened_results == 0 then
    return
  elseif #flattened_results == 1 then
    vim.lsp.util.jump_to_location(flattened_results[1])
  else
    local locations = vim.lsp.util.locations_to_items(flattened_results)
    pickers.new(opts, {
      prompt_title = 'LSP Definitions',
      finder = finders.new_table {
        results = locations,
        entry_maker = opts.entry_maker or make_entry.gen_from_quickfix(opts),
      },
      previewer = conf.qflist_previewer(opts),
      sorter = conf.generic_sorter(opts),
    }):find()
  end
```

Comment (jaywonchung) [reply lua/telescope/builtin/lsp.lua:72]: Just a small suggestion to avoid calling `vim.lsp.util.locations_to_items` in the common case :)

**Review by jaywonchung — APPROVED** [review]
Thank you for your great work!

I've tested the PR as is in my local machine and it works wonderfully, hence the approve. However I did leave one suggestion.

In addition, I think we should add the following item to `feature_map`:
```lua
["definitions"] = "goto_definition",
```

**Comment by jaywonchung** [comment]
Oh, and we should update `README.md`. Is this usually done in a separate PR?

**Comment by elianiva** [comment]
>  Is this usually done in a separate PR?

nope, we should do it in this PR :)
(and probably add a demo for the wiki, too)

**Comment by oberblastmeister** [comment]
I think for the demo someone else can add it for consistency. My vim doesn't look as nice and has a different colorscheme

**Comment by oberblastmeister** [comment]
also what is the feature map

**Comment by jaywonchung** [comment]
https://github.com/nvim-telescope/telescope.nvim/blob/1c80a92d4027c03b65cce1fe27ef664cff77186d/lua/telescope/builtin/lsp.lua#L245-L250

Here! It translates builtin function names to the corresponding LSP capability name.

File: lua/telescope/builtin/lsp.lua [hunk 2]
```diff
@@ -246,6 +276,7 @@ local feature_map = {
   ["code_actions"]      = "code_action",
   ["document_symbols"]  = "document_symbol",
   ["references"]        = "find_references",
+  ["definitions"]        = "definition",
```
Comment (fdschmidt93) [inline lua/telescope/builtin/lsp.lua:279]:
```suggestion
  ["definitions"]        = "goto_definition",
```
I think the corresponding LSP capability is slightly off. You can check them, for instance, with
```lua
print(vim.inspect(vim.lsp.buf_get_clients(0)[1].resolved_capabilities))
```
in which one of the table keys is `goto_definition`

Currently, it should not be able to go to a definition as the feature_map check always returns `false`.

Comment (jaywonchung) [reply lua/telescope/builtin/lsp.lua:279]: Yup! :+1::+1:

File: lua/telescope/builtin/lsp.lua [hunk 3]
```diff
@@ -246,6 +276,7 @@ local feature_map = {
   ["code_actions"]      = "code_action",
   ["document_symbols"]  = "document_symbol",
   ["references"]        = "find_references",
+  ["definitions"]        = "goto_definition",
```
Comment (jaywonchung) [inline lua/telescope/builtin/lsp.lua:279]:
```suggestion
  ["definitions"]       = "goto_definition",
```
Aesthetic purposes XD

**Review by Conni2461 — APPROVED** [review]
LGTM, haven't tested it tho (been pretty busy lately sorry :)

So if it works, feel free to squash and merge @oberblastmeister

**Comment by elianiva** [comment]
could someone tell me how to get multiple definition results? I was going to make the demo but I failed to get multiple results so it doesn't show the Telescope window

**Comment by jaywonchung** [comment]
```python
from typing import overload, Union

class KVStore:
    def __init__(self):
        self.store = dict()

    @overload
    def set(self, key: bytes, value: int) -> None:
        ...

    @overload
    def set(self, key: bytes, value: str) -> None:
        ...

    def set(self, key: bytes, value: Union[int, str]) -> None:
        self.store[key] = value


if __name__ == "__main__":
    store = KVStore()
    store.set(key=b"KEY", value=None)
```

with the cursor on `set` in the last line.

I use the `pyright` language server: https://github.com/microsoft/pyright

Relevant config:
```lua
if vim.fn.executable('pyright') == 1 then
  require'lspconfig'.pyright.setup{
    on_attach = require'completion'.on_attach,
    settings = {
      python = {
        analysis = {
          typeCheckingMode = "off",
          autoSearchPaths = true,
          useLibraryCodeForTypes = true,
        }
      }
    }
  }
  vim.cmd('autocmd FileType python setlocal omnifunc=v:lua.vim.lsp.omnifunc')
  vim.cmd('autocmd FileType python setlocal signcolumn=yes')
end
```

Screenshot of my editor:
스크린샷 2021-03-10 오후 3 14 23(2)

Hope this helps!

**Pull request merged**